File: 31. Next Permutation Medium/solution.ts
function nextPermutation(nums: number[]): void {
    let i = nums.length - 2;
    while (i >= 0 && nums[i] >= nums[i + 1]) i--;

    if (i >= 0) {
        let j = nums.length - 1;
        while (nums[j] <= nums[i]) j--;
        [nums[i], nums[j]] = [nums[j], nums[i]];
    }

    let [left, right] = [i + 1, nums.length - 1];

    while (left < right) {
        [nums[left], nums[right]] = [nums[right], nums[left]];
        left++;
        right--;
    }
}

let nums1 = [1,2,3];
nextPermutation(nums1);
console.log(nums1);

let nums2 = [3,2,1];
nextPermutation(nums2);
console.log(nums2);

let nums3 = [1,1,5];
nextPermutation(nums3);
console.log(nums3);

File: 31. Next Permutation Medium/solution.py
def next_permutation(nums: list[int]) -> None:
    i = len(nums) - 2
    while i >= 0 and nums[i] >= nums[i + 1]:
        i -= 1

    if i >= 0:
        j = len(nums) - 1
        while nums[j] <= nums[i]:
            j -= 1
        nums[i], nums[j] = nums[j], nums[i]
    
    left, right = i + 1, len(nums) - 1
    while left < right:
        nums[left], nums[right] = nums[right], nums[left]
        left += 1
        right -= 1

nums1 = [1,2,3]
next_permutation(nums1)
print(nums1)

nums2 = [3,2,1]
next_permutation(nums2)
print(nums2)

nums3 = [1,1,5]
next_permutation(nums3)
print(nums3)